### PR TITLE
Topbar

### PR DESCRIPTION
**Pull request by Nosoyleon**
## Summary
1. Structuring AthorizedRoute
2. Creating Topbar Component
3. Creating Empty Profile view
4. Handling log out action
5. Using connected-react-router

## Screenshots
image

## Trello Card
https://trello.com/c/2dt7vTOS/25-routing-ii-topbar

### CODE REVIEW

File: react/training-project/src/app/index.js [hunk 1]
```diff
@@ -17,16 +19,20 @@ class App extends Component {
           <img src={logo} className={style.appLogo} alt="logo" />
           <h1 className={style.appTitle}>{this.APP_TITLE}</h1>
         </header>
-        <Router>
+        <ConnectedRouter history={this.props.history}>
           <Switch>
             <Route path="/login" component={Login} />
-            <Route path="/" exact component={Game} />
+            <Route path="/app" component={AuthorizedRoute} />
```
Comment (BJordi) [inline react/training-project/src/app/index.js]: Create a constants file and add there all the paths.

File: react/training-project/src/app/screens/AuthorizedRoute/components/Topbar/index.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import { NavLink } from 'react-router-dom';
+import PropTypes from 'prop-types';
+
+import style from './styles.scss';
+
+function Topbar({ logOut }) {
+  return (
+    <nav className={style.navbar}>
+      <NavLink className={style.navbarItem} activeClassName={style.active} to="/app/game">
+        Play
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/components/Topbar/index.js]: Create a constants file in this component folder and add there all there literals.

File: react/training-project/src/app/screens/AuthorizedRoute/components/Topbar/index.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      </NavLink>
+      <NavLink className={style.navbarItem} activeClassName={style.active} to="/app/Profile">
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/components/Topbar/index.js]: Paths -> to the routes constants file

File: react/training-project/src/app/screens/AuthorizedRoute/components/Topbar/styles.scss [hunk 4]
```diff
@@ -0,0 +1,34 @@
+@import 'src/scss/colors';
+@import 'src/scss/fixed-sizes';
+
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/components/Topbar/styles.scss:3]: Remove two of these empty lines.

File: react/training-project/src/app/screens/AuthorizedRoute/index.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import Game from './screens/Game';
+import Profile from './screens/Profile';
+
+class AuthorizedRoute extends Component {
+  handleLogOut = () => this.props.logOut(this.props.token);
+
+  render() {
+    const { match } = this.props;
+    return !this.props.token ? (
+      <Redirect to="/login" />
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/index.js]: Same here.

File: react/training-project/src/app/screens/AuthorizedRoute/index.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+AuthorizedRoute.propTypes = {
+  match: PropTypes.objectOf(PropTypes.any).isRequired,
+  token: PropTypes.string,
+  logOut: PropTypes.func.isRequired
+};
+
+const mapStateToProps = state => ({
+  state,
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/index.js]: Why is this line here? :thinking:

File: react/training-project/src/app/screens/AuthorizedRoute/index.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  render() {
+    const { match } = this.props;
+    return !this.props.token ? (
+      <Redirect to="/login" />
+    ) : (
+      <Fragment>
+        <Topbar logOut={this.handleLogOut} />
+        <Switch>
+          <Route path={`${match.path}/game`} component={Game} />
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/index.js]: I would use constants instead of interpolating.

**Review by BJordi — CHANGES_REQUESTED** [review]

File: react/training-project/src/app/screens/AuthorizedRoute/screens/Game/index.js [hunk 8]
```diff
@@ -51,9 +38,10 @@ Game.propTypes = {
 };
 
 const mapStateToProps = state => ({
-  history: state.history,
-  stepNumber: state.stepNumber,
-  xIsNext: state.xIsNext
+  state,
```
Comment (BJordi) [inline react/training-project/src/app/screens/AuthorizedRoute/screens/Game/index.js]: Why are you adding state here? :thinking:

File: react/training-project/src/app/screens/Login/index.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import { Redirect } from 'react-router-dom';
+import PropTypes from 'prop-types';
+
+import actionCreators from '~redux/Auth/actions';
+
+import LoginForm from './layout';
+
+/* eslint-disable no-func-assign */
+class LoginPage extends Component {
+  handleSubmit = values => {
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/index.js]: Try to be consistent. For one-line methods you're using two syntax; sometimes you use this one:
```
handleSubmit = values => {
  this.props.logIn(values);
}
```
And sometimes you use this one:
`
handleSubmit = values => this.props.logIn(values);
`

File: react/training-project/src/app/screens/Login/index.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+  render() {
+    return !this.props.token ? (
+      <LoginForm
+        processing={this.props.processing}
+        loginFail={this.props.loginFail}
+        onSubmit={this.handleSubmit}
+      />
+    ) : (
+      <Redirect to="/app/game" />
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/index.js:22]: Use constants for paths. Create a constants.js file in the Routes folder.

File: react/training-project/src/app/screens/Login/index.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+LoginPage.propTypes = {
+  logIn: PropTypes.func.isRequired,
+  loginFail: PropTypes.bool.isRequired,
+  token: PropTypes.string,
+  processing: PropTypes.bool.isRequired
+};
+
+const mapStateToProps = state => ({
+  state,
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/index.js]: Same here. Why are you adding this?

File: react/training-project/src/app/screens/Login/layout.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+          validate={[required, minLength]}
+          component={Input}
+          type="password"
+          placeholder="Password"
+        />
+        <button className={style.loginButton} type="submit">
+          {processing ? 'Processing' : 'Log in'}
+        </button>
+      </form>
+      {loginFail && <Message>Username or password invalid!</Message>}
```
Comment (BJordi) [inline react/training-project/src/app/screens/Login/layout.js:37]: Use constants

File: react/training-project/src/redux/Auth/actions.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+  logOutSuccess: () => ({
+    type: actions.LOG_OUT_SUCCESS,
+    payload: null
+  })
+};
+
+const actionCreators = {
+  logIn: ({ username, password }) => async dispatch => {
+    dispatch(privateActionsCreators.logInProcessing(true));
```
Comment (BJordi) [inline react/training-project/src/redux/Auth/actions.js:48]: You can move the content of logInProcessing here. But if you like it this way, it's ok too! :wink:

File: react/training-project/src/redux/Auth/actions.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    const response = await AuthService.logIn({ username, password });
+
+    if (response.ok && response.data.length > 0) {
```
Comment (BJordi) [inline react/training-project/src/redux/Auth/actions.js:51]: data is an array? :thinking: If it is, then, you should check the existence of `response.data`. I know that if `response.ok` is true, `response.data` should exists, but it's always better to check this things and prevent some errors.

File: react/training-project/src/redux/Auth/actions.js [hunk 15]
```diff
@@ -0,0 +1,74 @@
+import AuthService from '../../services/AuthService';
+
+export const actions = {
+  LOG_IN: '@@AUTH/LOG_IN',
+  LOG_IN_SUCCESS: '@@AUTH/LOG_IN_SUCCESS',
+  LOG_IN_FAIL: '@@AUTH/LOG_IN_FAIL',
+  LOG_IN_PROCESSING: '@AUTH/LOG_IN_PROCESSING',
+  LOG_OUT: '@@AUTH/V',
```
Comment (BJordi) [inline react/training-project/src/redux/Auth/actions.js]: V? :thinking:

File: react/training-project/src/redux/Auth/actions.js [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  },
+
+  logOut: token => async dispatch => {
+    const response = await AuthService.getToken(token);
+
+    if (response.ok && response.data.length > 0) {
+      response.data.forEach(async element => {
+        const deletedToken = await AuthService.logOut(element.id);
+        if (!deletedToken.ok) return { type: actions.LOG_OUT_FAIL };
```
Comment (BJordi) [inline react/training-project/src/redux/Auth/actions.js]: Not dispatching?

**Review by BJordi — CHANGES_REQUESTED** [review]

File: react/training-project/src/app/screens/Login/index.js (hunk 11 above)
Comment (Nosoyleon) [reply react/training-project/src/app/screens/Login/index.js]: already solved in PR19

File: react/training-project/src/app/screens/Login/layout.js (hunk 12 above)
Comment (Nosoyleon) [reply react/training-project/src/app/screens/Login/layout.js:37]: Already solved in PR19

File: react/training-project/src/app/screens/Login/index.js (hunk 10 above)
Comment (Nosoyleon) [reply react/training-project/src/app/screens/Login/index.js:22]: already solved in PR19

**Review by BJordi — APPROVED** [review]
